operator/src/command_utils.rs

```diff
@@ -0,0 +1,306 @@
+            message: Some(CommandStatusMessage::Running),
+        };
+
+        self.write_status(&client, new_status).await
+    }
+
+    /// Implementation of behavior when commands are finished (at the end of reconcile)
+    ///
+    /// # Arguments
+    /// * `client` - Kubernetes client
+    ///
+    pub async fn process_command_finalize(&self, client: &Client) -> OperatorResult<bool> {
+        let new_status = &CommandStatus {
+            started_at: None,
+            finished_at: get_time(),
+            message: Some(CommandStatusMessage::Finished),
+        };
+
+        self.write_status(&client, new_status).await
+    }
+}
+
+/// Initialize all comments with a status message if no status available yet. Returns true
+/// if any status has been updated.
+///
+/// # Arguments
+/// * `client` - Kubernetes client
+/// * `commands` - Available commands
+///
+pub async fn init_commands(client: &Client, commands: &[CommandType]) -> OperatorResult<bool> {
+    let mut changed = false;
+    for command in commands {
+        // patch empty status first
+        if command.get_status().is_none() {
+            let new_status = CommandStatus {
+                started_at: None,
+                finished_at: None,
+                message: Some(CommandStatusMessage::Enqueued),
+            };
+
+            command.write_status(client, &new_status).await?;
+
+            changed = true;
+        }
+    }
+    Ok(changed)
+}
+
+/// Collect all different commands in one vector.
+///
+/// # Arguments
+/// * `client` - Kubernetes client
+///
+pub async fn collect_commands(client: &Client) -> OperatorResult<Vec<CommandType>> {
+    let mut all_commands = vec![];
+    let mut restart_commands: Vec<Restart> =
+        stackable_operator::command_controller::list_commands::<Restart>(client).await?;
+
+    let mut start_commands: Vec<Start> =
+        stackable_operator::command_controller::list_commands::<Start>(client).await?;
+
+    let mut stop_commands: Vec<Stop> =
+        stackable_operator::command_controller::list_commands::<Stop>(client).await?;
+
+    while let Some(cmd) = restart_commands.pop() {
+        all_commands.push(CommandType::Restart(cmd));
+    }
+
+    while let Some(cmd) = start_commands.pop() {
+        all_commands.push(CommandType::Start(cmd));
+    }
+
+    while let Some(cmd) = stop_commands.pop() {
+        all_commands.push(CommandType::Stop(cmd));
+    }
+
+    Ok(all_commands)
+}
+
+/// Retrieve the current command depending on its status. We search for "enqueued" commands which
+/// have no startedAt and finishedAt timestamps, as well as started/running commands that have
+/// a startedAt but no finishedAt timestamp. We ignore finished commands (startedAt, finishedAt
+/// are set).
+///
+/// # Arguments
+/// * `command_information` - CommandInformation containing a list of all commands
+///
+pub fn get_current_command(
+    command_information: &Option<CommandInformation>,
+) -> Result<Option<&CommandType>, Error> {
```

Comment:
This is a genuine question: If a condition can only happen due to a programming error would it make sense to panic instead? @soenkeliebau @siegfriedweber 

What happens now is that we'll requeue and due to the same bug we'll hit the same error and will requeue and so on.... people might not even notice. If however we crash someone could see it in their monitoring and see that something's wrong and then fix the object in question manually.

I'm not sure...


Also Result -> OperatorResult

Reply:
If its only a due a programming error i agree. But especially with the custom resources, they could be manipulated by a  mean cluster troll ;-)
The question is how to deal with that? Delete it? Ignore that?

Reply:
If something can really only happen due to a programming error I'd say it is fine to panic. I believe we have that in a couple of places already and encourage people to open a ticket in that case.
If we continue down that road I'd like to track these occurrences somewhere in the docs, maybe even with a reference that customers can use when opening a ticket.